File: app/api/categories.py
from flask import Blueprint, request, jsonify
from app import db
from models.usermodel import Category  # Adjust import path based on your structure

category_bp = Blueprint('category', __name__, url_prefix='/categories')

@category_bp.route('/', methods=['GET'])
def get_categories():
    categories = Category.query.all()
    categories_list = [{'id': category.id, 'name': category.name} for category in categories]
    return jsonify(categories_list)

@category_bp.route('/create', methods=['POST'])
def create_category():
    name = request.form.get('name')

    if not name:
        return jsonify({"error": "Missing required field: name"}), 400

    category = Category(name=name)
    db.session.add(category)
    db.session.commit()

    return jsonify({"message": "Category created successfully", "category_id": category.id}), 201


@category_bp.route('/<int:id>/edit', methods=['PUT'])
def edit_category(id):
    category = Category.query.get_or_404(id)
    name = request.json.get('name')

    if not name:
        return jsonify({"error": "Missing required field: name"}), 400

    category.name = name
    db.session.commit()

    return jsonify({"message": "Category updated successfully"}), 200


@category_bp.route('/<int:id>/delete', methods=['DELETE'])
def delete_category(id):
    category = Category.query.get_or_404(id)
    db.session.delete(category)
    db.session.commit()

    return jsonify({"message": "Category deleted successfully"}), 200

File: app/static/js/categories.js
function createCategory() {
    // Get the category name from the input field
    var categoryName = document.getElementById('categoryName').value;

    // Validate if category name is provided
    if (!categoryName) {
        alert('Please provide a category name');
        return;
    }

    // Prepare data to send to the server
    var formData = new FormData();
    formData.append('name', categoryName);

    // Send POST request to the server
    $.ajax({
        url: '/categories/create',
        type: 'POST',
        data: formData,
        processData: false,
        contentType: false,
        success: function(response) {
            toastr.success('Category created successfully! Category ID: ' + response.category_id);
            // Optionally, you can reset the form here
            form.reset();
        },
        error: function(xhr, status, error) {
            console.log(xhr); // Log the entire xhr object to see its structure and contents
            var errorMessage = xhr.status + ': ' + xhr.statusText;
            toastr.error('Error - ' + errorMessage);
        }
    });
}
// Function to display categories
function displayCategories() {
    fetch('/categories') // Assuming this endpoint returns category data
        .then(response => response.json())
        .then(categories => {
            console.log(categories); // Log the categories data
            console.log("done");
            const tableBody = document.querySelector('#categoryTable tbody');
            tableBody.innerHTML = ''; // Clear existing table rows

            categories.forEach(category => {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${category.id}</td>
                    <td>${category.name}</td>
                    <td>
                        <button onclick="editCategory(${category.id})">Edit</button>
                        <button onclick="deleteCategory(${category.id})">Delete</button>
                    </td>
                `;
                tableBody.appendChild(row);
            });
        })
        .catch(error => {
            console.error('Error fetching categories:', error);
        });
}


// Function to call displayCategories() when the page loads
window.onload = displayCategories;

